libfwupdplugin/fu-progress.c

```diff
@@ -0,0 +1,714 @@
+ *
+ * Returns if the profile is enabled for this progress.
+ *
+ * Return value: if profiling should be enabled
+ *
+ * Since: 1.7.0
+ **/
+static gboolean
+fu_progress_get_profile(FuProgress *self)
+{
+	FuProgressPrivate *priv = GET_PRIVATE(self);
+	g_return_val_if_fail(FU_IS_PROGRESS(self), FALSE);
+	return priv->profile;
+}
+
+/**
+ * fu_progress_reset:
+ * @self: A #FuProgress
+ *
+ * Resets the #FuProgress object to unset
+ *
+ * Since: 1.7.0
+ **/
+void
+fu_progress_reset(FuProgress *self)
+{
+	FuProgressPrivate *priv = GET_PRIVATE(self);
+	g_return_if_fail(FU_IS_PROGRESS(self));
+
+	/* reset values */
+	priv->step_max = 0;
+	priv->step_now = 0;
+	priv->percentage = 0;
+
+	/* only use the timer if profiling; it's expensive */
+	if (priv->profile)
+		g_timer_start(priv->timer);
+
+	/* disconnect client */
+	if (priv->percentage_child_id != 0) {
+		g_signal_handler_disconnect(priv->child, priv->percentage_child_id);
+		priv->percentage_child_id = 0;
+	}
+
+	/* unref child */
+	g_clear_object(&priv->child);
+
+	/* no more step data */
+	g_clear_pointer(&priv->step_data, g_free);
+	g_clear_pointer(&priv->step_profile, g_free);
+}
+
+/**
+ * fu_progress_set_steps_full:
+ * @self: A #FuProgress
+ * @id: the code location
+ * @step_max: The number of sub-tasks in this progress, can be 0
+ *
+ * Sets the number of sub-tasks, i.e. how many times the fu_progress_step_done()
+ * function will be called in the loop.
+ *
+ * The function will immediately return when the number of step_max is 0
+ * or if fu_progress_set_enabled(FALSE) was previously called.
+ *
+ * Since: 1.7.0
+ **/
+void
+fu_progress_set_steps_full(FuProgress *self, const gchar *id, guint step_max)
+{
+	FuProgressPrivate *priv = GET_PRIVATE(self);
+	g_return_if_fail(FU_IS_PROGRESS(self));
+
+	/* nothing to do for 0 step_max */
+	if (step_max == 0)
+		return;
+
+	/* did we call done on a self that did not have a size set? */
+	if (priv->step_max != 0) {
+		g_autoptr(GString) str = g_string_new(NULL);
+		fu_progress_build_parent_chain(self, str, 0);
+		g_warning("step_max already set to %u, can't set %u! [%s]: %s",
+			  priv->step_max,
+			  step_max,
+			  id,
+			  str->str);
+		return;
+	}
+
+	/* set id */
+	g_free(priv->id);
+	priv->id = g_strdup(id);
+
+	/* only use the timer if profiling; it's expensive */
+	if (priv->profile)
+		g_timer_start(priv->timer);
+
+	/* imply reset */
+	fu_progress_reset(self);
+
+	/* set step_max */
+	priv->step_max = step_max;
+
+	/* global share just got smaller */
+	priv->global_share /= step_max;
+}
+
+/**
+ * fu_progress_set_custom_steps_full:
+ * @self: A #FuProgress
+ * @id: the code location
+ * @value: A step weighting variable argument array
+ *
+ * This sets the step weighting, which you will want to do if one action
+ * will take a bigger chunk of time than another.
+ *
+ * All the values must add up to 100, and the list must end with -1.
+ * Do not use this function directly, instead use the fu_progress_set_custom_steps() macro.
```

Comment:
should it be in a private header to prevent that?

Reply:
Hmm, this didn't work as well as I hoped. As it's a pre-processor directive (as we need G_STRLOC) you need to be able to find the `fu_progress_set_custom_steps_full` prototype when compiling the code. The alternative is we do something like:

	fu_progress_set_custom_steps(progress, G_STRLOC, 19 /* erase */, 45 /* write */, 36 /* read */, -1);

But that's kinda _nrrrgh_. Other ideas welcome. The reason we save the G_STRLOC is so we can tell the user _which_ `fu_progress_set_custom_steps()` needs fixing up when profiling, or when the user used to many `fu_progress_step_done()`'s

Reply:
Maybe add individual steps (`fu_progress_add_step`)

Reply:
I was _dreaming_ about this, so it's probably about time to go on holiday. Anyway. Where I think this might make sense is doing something like:

	fu_progress_set_custom_steps(progress, 10 /* erase */, 70 /* write */, 20 /* verify */, -1);

	/* erase phase */
	fu_device_set_status(device,FWUPD_STATUS_DEVICE_ERASE)
	...
	fu_progress_step_done(progress);

	/* write phase */
	fu_device_set_status(device,FWUPD_STATUS_DEVICE_WRITE)
	...
	fu_progress_step_done(progress);

	/* verify phase */
	fu_device_set_status(device,FWUPD_STATUS_DEVICE_VERIFY)
	...
	fu_progress_step_done(progress);

and instead have someting like:

	fu_progress_set_id(progress, G_STRLOC);
	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_ERASE, 1);
	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 7);
	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_VERIFY, 2);

	/* erase phase */
	...
	fu_progress_step_done(progress);

	/* write phase */
	...
	fu_progress_step_done(progress);

	/* verify phase */
	...
	fu_progress_step_done(progress);

this means we decouple the "must add up to 100%" thing and the numbers become just scaled factors. It also means the device state gets set automatically at each stage.

Reply:
I added this as a fixup commit -- I think it's much nicer. Comments welcome.